docs/architecture/adr-021-protobuf-query-encoding.md

```diff
@@ -0,0 +1,262 @@
+using the `WithValue` and `Value` methods, the SDK should provide a function
+`sdk.UnwrapSDKContext` to retrieve the `sdk.Context` from the provided
+`context.Context`.
+
+An example implementation of `QueryBalance` for the bank module as above would
+look something like:
+
+```go
+type Querier struct {
+	Keeper
+}
+
+func (q Querier) QueryBalance(ctx context.Context, params *types.QueryBalanceParams) (*sdk.Coin, error) {
+	balance := q.GetBalance(sdk.UnwrapSDKContext(ctx), params.Address, params.Denom)
+	return &balance, nil
+}
+```
+
+### Custom Query Registration and Routing
+
+Query server implementations as above would be registered with `AppModule`s using
+a new method `RegisterQueryServer(grpc.Server)` which could be implemented simply
+as below:
+
+```go
+// x/bank/module.go
+func (am AppModule) RegisterQueryServer(server grpc.Server) {
+	types.RegisterQueryServer(server, keeper.Querier{am.keeper})
+}
+```
+
+Underneath the hood, a new method `RegisterService(sd *grpc.ServiceDesc, handler interface{})`
+will be added to the existing `baseapp.QueryRouter` to add the queries to the custom
+query routing table (with the routing method being described below).
+The signature for this method matches the existing
+`RegisterServer` method on the GRPC `Server` type where `handler` is the custom
+query server implementation described above.
+
+
+GRPC-like requests are routed by the service name (ex. `cosmos_sdk.x.bank.v1.Query`)
+and method name (ex. `QueryBalance`) combined with `/`s to form a full
+method name (ex. `/cosmos_sdk.x.bank.v1.Query/QueryBalance`). This gets translated
+into an ABCI query as `custom/cosmos_sdk.x.bank.v1.Query/QueryBalance`. Service handlers
+registered with `QueryRouter.RegisterService` will be routed this way.
+
+Beyond the method name, GRPC requests carry a protobuf encoded payload, which maps naturally
+to `RequestQuery.Data`, and receive a protobuf encoded response or error. Thus
+there is a quite natural mapping of GRPC-like rpc methods to the existing
+`sdk.Query` and `QueryRouter` infrastructure.
+
+This basic specification allows us to reuse protocol buffer `service` definitions
+for ABCI custom queries substantially reducing the need for manual decoding and
+encoding in query methods. 
+
+### GRPC Protocol Support
+
+In addition to providing an ABCI query pathway, we can easily provide a GRPC
+proxy server that routes requests in the GRPC protocol to ABCI query requests
+under the hood. In this way, clients could use their host languages' existing
+GRPC implementations to make direct queries against Cosmos SDK app's using
+these `service` definitions. In order for this server to work, the `QueryRouter`
+on `BaseApp` will need to expose the service handlers registered with
+`QueryRouter.RegisterService` to the proxy server implementation. Nodes could
+launch the proxy server on a separate port in the same process as the ABCI app
+with a command-line flag.
```

Comment:
Do operators have to launch and manage two separate processes now? Or can this be automatically handled via a single command?

Reply:
They could launch the gRPC proxy, REST proxy, and ABCI app all within one process with a single command. That's what I have in mind here. There could also be commands to do it via separate processes.

Reply:
What is the REST proxy? I presume you mean REST server as it stands now? If so, that should be started and handled separately as it is now as well.

Why not have a single command that starts the REST server and gRPC proxy as the same time?

Reply:
Well there's still the REST server as it stands now. That wouldn't be touched by any of this although I imagine it will become less relevant as a result.

 What I mean by REST proxy is grpc-gateway. So a single command that starts the gRPC proxy + grpc-gateway in the same process (or separate process) as the ABCI app makes sense.

Reply:
So in that case, the REST server and command become obsolete if I understand correctly?

Reply:
Yes, eventually. Maybe it's still needed for transactions? Although maybe that also becomes obsolete by actually supporting protobuf for transactions??

Reply:
Yikes, that sounds like a poor operator UX. I wouldn't want to spin up a server just to generate txs. It should be supported by this new gRPC work. i.e. POST to create a tx. Is this possible?

Reply:
So how REST tx's work is something I'm not super familiar with.

Is the idea that someone can spin up a local REST client with a local keybase and submit transactions through JSON as opposed to the CLI? Or is there some other use case?

If that's so then it seems like the need for this would be mostly obviated by protobuf to begin with. But if something similar were needed, sure you could use gRPC to do the same thing.

Reply:
No, there is no Keybase -- users/client's dont sign anything. The tx side of the REST server does two things:

1. Generate unsigned txs (exactly the same as the CLI -- see my bank PR for ref).
2. Broadcast an alread signed tx.

I would imagine gRPC can simply have POST methods to generate unsigned txs, right?

Reply:
Well just to be clear gRPC itself doesn't differentiate between GET and POST, grpc-gateway does when translating gRPC -> REST. But sure, if a gRPC call is needed to generate an unsigned tx of course.

I guess I was just questioning whether that is still needed given that protobuf should make it easier to generate unsigned tx's anyway.

If it is still needed, we can either keep the existing REST routes or migrate them to gRPC/grpc-gateway in which case a module would probably expose a `Tx` service. 

I do want to note that either way (existing REST or gRPC) we'll encounter the same complexity we have everywhere else with app-level `oneof`s for interfaces. This also will apply to `tx` CLI. So limiting the number of places that complexity needs to arise is probably best. Thus if clients can just consume `.proto` files to generate unsigned tx's, maybe that's enough...?

Reply:
I would also propose we address REST tx's in a separate issue if what is here for queries is fine.

Reply:
> If it is still needed, we can either keep the existing REST routes or migrate them to gRPC/grpc-gateway in which case a module would probably expose a Tx service.

Yes this is what Im proposing.

>  This also will apply to tx CLI.

This is already done. See my bank PR. These components will need to be passed interfaces that know how to construct the necessary messages.

Reply:
I would also state that is feature is heavily relied upon by many clients. I would highly recommend the work be done in tandem.

Reply:
> Yes this is what Im proposing. The new Querier/gRPC Service that each module now implements will have the functionality to generate unsigned txs.

Okay, is it fine to address unsigned tx's separately from this ADR?

Reply:
If it is heavily depended upon by clients then isn't the best approach to stick with the existing REST tx handlers because that will be the path of least change? Also do they depend on it because of a lack of amino implementations which protobuf will solve? Or some other reason? My preference would be to keep this ADR to queries and address REST tx separately.

Reply:
> If it is heavily depended upon by clients then isn't the best approach to stick with the existing REST tx handlers because that will be the path of least change?

I mentioned earlier this UX sounds like a poor design. Ideally, they would not have to think, "ohhh, the binary daemon now serves queries, but I need to spin up another daemon just to serve tx generation..."

> Also do they depend on it because of a lack of amino implementations which protobuf will solve? Or some other reason? 

I imagine it varies, but most likely because it's extremely easy and effortless.

>  preference would be to keep this ADR to queries and address REST tx separately either way - this mostly touches upon the querier and not REST (it just happens to also cover REST because of grpc-gateway).

That's fine, but just know I wouldn't consider this feature complete until it can also address tx generation. So keep that in mind in your design.

Reply:
> I mentioned earlier this UX sounds like a poor design. Ideally, they would not have to think, "ohhh, the binary daemon now serves queries, but I need to spin up another daemon just to serve tx generation..."

I'm pretty sure we can figure out a decent UX for a single daemon with multiple ports exposed.

But I do think there's a bunch of other things to discuss and would prefer to have a separate ADR for tx generation. For instance, there's the fact that we have Tx defined in the app-level proto file. Anyway, now that you've brought it up I'll definitely be thinking about it.